rust/Dockerfile

```diff
@@ -1,65 +1,64 @@
-# syntax=docker/dockerfile:experimental
+# syntax=docker/dockerfile:1.4
 
-FROM rust:1.80.1 as builder
-WORKDIR /usr/src
-
-# 1a: Prepare for static linking
+# -------- Base Image with Tools --------
+FROM rust:1.80.1 AS base
 RUN apt-get update && \
-    apt-get dist-upgrade -y && \
     apt-get install -y musl-tools clang && \
-    rustup target add x86_64-unknown-linux-musl
+    rustup target add x86_64-unknown-linux-musl && \
+    cargo install --locked sccache
+
+ENV RUSTC_WRAPPER=sccache
+ENV SCCACHE_DIR=/sccache
 
-RUN mkdir -p rust/main
-RUN mkdir -p rust/sealevel
+# -------- Builder Stage --------
+FROM base AS builder
```

Comment:
what's this?

Reply:
previously we just had the build step and then the runtime image that gets created. now we're separating into a `base` image, and then later a `builder` image which builds from this base image. This is so that we have better separation/caching of the parts that don't change frequently (toolchain & setup) vs things that are more likely to change (the actual rust compilation)

Reply:
oh i see it here: `FROM rust:1.80.1 AS base`

that's sick, seeing here: https://docs.docker.com/build/building/multi-stage/#use-multi-stage-builds

but yeah as Mo mentioned, would be good to add some docs in the PR descriptions since none of us are docker experts so it's hard to tell what's going on

Reply:
added comments throughout the file to explain what it's doing better - also documented existing steps for future readers